crates/ruff_python_formatter/src/expression/string/docstring.rs

```diff
@@ -702,19 +784,336 @@ impl<'src> CodeExampleDoctest<'src> {
+            // the `code-block` directive. Because of how we handle it here, the above
+            // is not treated as a code snippet.
+            //
+            // ... but it
+            return None;
+        }
+        // At this point, we know we didn't find a code block, so the only
+        // thing we can hope for is a literal block which must end with a `::`.
+        if !rest.trim_end().ends_with("::") {
+            return None;
+        }
+        Some(CodeExampleRst {
+            lines: vec![],
+            opening_indent: indentation_length(opening_indent),
+            min_indent: None,
+            is_directive: false,
+        })
+    }
+
+    /// Attempts to create a new reStructuredText code example from a
+    /// `code-block` or `sourcecode` directive. If one couldn't be found, then
+    /// `None` is returned.
+    fn new_code_block(original: InputDocstringLine<'src>) -> Option<CodeExampleRst> {
+        // This regex attempts to parse the start of a reStructuredText code
+        // block [directive]. From the reStructuredText spec:
+        //
+        // > Directives are indicated by an explicit markup start (".. ")
+        // > followed by the directive type, two colons, and whitespace
+        // > (together called the "directive marker"). Directive types
+        // > are case-insensitive single words (alphanumerics plus
+        // > isolated internal hyphens, underscores, plus signs, colons,
+        // > and periods; no whitespace).
+        //
+        // The language names matched here (e.g., `python` or `py`) are taken
+        // from the [Pygments lexer names], which is referenced from the docs
+        // for the [code-block] directive.
+        //
+        // [directives]: https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#directives
+        // [Pygments lexer names]: https://pygments.org/docs/lexers/
+        // [code-block]: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-code-block
+        static DIRECTIVE_START: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(
+                r"(?m)^\s*\.\. \s*(?i:code-block|sourcecode)::\s*(?i:python|py|python3|py3)$",
+            )
+            .unwrap()
+        });
+        if !DIRECTIVE_START.is_match(original.line) {
+            return None;
+        }
+        Some(CodeExampleRst {
+            lines: vec![],
+            opening_indent: indentation_length(original.line),
+            min_indent: None,
+            is_directive: true,
+        })
+    }
+
+    /// Returns the code collected in this example as a sequence of lines.
+    ///
+    /// The lines returned have the minimum indentation stripped from their
+    /// prefix. Based on the definition of minimum indentation, this implies
+    /// there is at least one line in the slice returned with no whitespace
+    /// prefix.
+    fn code(&mut self) -> &[CodeExampleLine<'src>] {
```

Comment:
I find it unexpected that a function called `code` mutates the lines. It may take me a while to figure out where `self.lines` gets changed due to it. Could we trim the lines already when adding it or is this not possible because we don't know the `min_indent` yet?

Reply:
Correct. We don't know the `min_indent`.

I agree that the naming here is not ideal. I don't like it either. I would have liked to define an `into_code` that consumes `self` and spits back the code, but then you lose other parts of `self` that are needed to print the reformatted lines. I could also define a `take_code` instead that plucks the code out of an `&mut self`, which might make the mutation a little less surprising, but then you're left with an empty code example whose other state is still useful. Which is weird. Another choice would be to add a new method, something like, `indent_lines` that does the mutation and then provide the `code` method as a simple accessor. I probably like that approach the best, but it means you can forget to run `indent_lines` and get a bug.

The benefit of this approach is that callers can't get it wrong, which is ultimately what I ended up favoring.

Still, the name isn't great. I changed it to `indented_code` and tweaked the docs a touch. Still not great, but maybe a little better.